src/common/hashes/shamap.ts

```diff
@@ -0,0 +1,263 @@
+import {HASH_INNER_NODE, HASH_INNER_NODE_V2, HASH_TX_ID, HASH_TX_NODE, HASH_LEAF_NODE} from './hashprefix'
+import hash from './sha512hash'
+const HEX_ZERO = '0000000000000000000000000000000000000000000000000000000000000000'
+export const TYPE_INNER = 1
+export const TYPE_TRANSACTION_NM = 2
+export const TYPE_TRANSACTION_MD = 3
+export const TYPE_ACCOUNT_STATE = 4
+
+/**
+ * Abstract class representing a node in a SHAMap tree.
+ *
+ * Can be either SHAMapTreeNodeInner, SHAMapTreeNodeInnerV2
+ * or SHAMapTreeNodeLeaf.
+ *
+ * @class
+ */
+export abstract class Node {
+  public constructor() {}
+  public add_item(_tag: string, _node: Node): void {
+    throw new Error('Called unimplemented virtual method SHAMapTreeNode#add_item.')
+  }
+  public get hash(): string|void {
+    throw new Error('Called unimplemented virtual method SHAMapTreeNode#hash.');
+  }
+}
+
+/**
+ * Inner (non-leaf) node in a SHAMap tree.
+ * @param {Number} depth (i.e. how many parent inner nodes)
+ * @class
+ */
+export class InnerNode extends Node {
+  public leaves: object
+  public type: 1 | 2 | 3 |4
+  public depth: number
+  public empty: boolean
+  public constructor(depth: number = 0) {
+    super()
+    this.leaves = {}
+    this.type = TYPE_INNER
+    this.depth = depth
+    this.empty = true
+  }
+  /**
+   * @param {String} tag (equates to a ledger entry `index`)
+   * @param {Node} node (to add)
```

Comment:
OOC: why do you surround args in parenthesis?

Reply:
This was a remnant from the ripple-hashes documentation. I didn't modify the docs.